sshd-core/src/main/java/org/apache/sshd/common/config/keys/OpenSSHKeyUtils.java

```diff
@@ -0,0 +1,150 @@
+
+    private static final AtomicReference<Factory<? extends Digest>> DEFAULT_DIGEST_HOLDER =
+            new AtomicReference<Factory<? extends Digest>>(DEFAULT_FINGERPRINT_DIGEST_FACTORY);
+
+    private OpenSSHKeyUtils() {
+        throw new UnsupportedOperationException("No instance");
+    }
+
+    /**
+     * @return The default {@link Factory} of {@link Digest}s used
+     * by the {@link #getFingerPrint(PublicKey)} and {@link #getFingerPrint(String)}
+     * methods
+     * @see #setDefaultFingerPrintFactory(Factory)
+     */
+    public static Factory<? extends Digest> getDefaultFingerPrintFactory() {
+        return DEFAULT_DIGEST_HOLDER.get();
+    }
+
+    /**
+     * @param f The {@link Factory} of {@link Digest}s to be used - may
+     *          not be {@code null}
+     */
+    public static void setDefaultFingerPrintFactory(Factory<? extends Digest> f) {
+        DEFAULT_DIGEST_HOLDER.set(ValidateUtils.checkNotNull(f, "No digest factory"));
+    }
+
+    /**
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * <B>Note:</B> if exception encountered then returns the exception's simple class name
+     * @see #getFingerPrint(Factory, PublicKey)
+     */
+    public static String getFingerPrint(PublicKey key) throws Exception {
+        return getFingerPrint(getDefaultFingerPrintFactory(), key);
+    }
+
+    /**
+     * @param f   The {@link Factory} to create the {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * <B>Note:</B> if exception encountered then returns the exception's simple class name
+     * @see #getFingerPrint(Digest, PublicKey)
+     */
+    public static String getFingerPrint(Factory<? extends Digest> f, PublicKey key) throws Exception {
+        return getFingerPrint(f.create(), key);
+    }
+
+    /**
+     * @param d   The {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * @throws Exception if cannot create fingerprint.
+     */
+    public static String getFingerPrint(Digest d, PublicKey key) throws Exception {
+        if (key == null) {
+            return null;
+        }
+
+        Buffer buffer = new ByteArrayBuffer();
+        buffer.putRawPublicKey(key);
+
+        d.init();
+        d.update(buffer.array(), 0, buffer.wpos());
+
+        byte[] data = d.digest();
+
+        if ("MD5".equals(d.getAlgorithm())) {
+            return "MD5:" + BufferUtils.printHex(':', data);
+        } else {
+            return d.getAlgorithm().replace("-", "").toUpperCase() + ":" + Base64.encodeToString(data).replaceAll("=", "");
+        }
+    }
+
+    /**
+     * @param expected The expected fingerprint.
+     * @param key the public key - ignored if {@code null}
```

Comment:
If {@code null} / empty then returns a failure indication with the default fingerprint. If no digest specified the the default digest is used.